src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,725 @@
+                    events.push(Event {
+                        token: pending_wake_token,
+                        events: notified_events,
+                    });
+                }
+
+                for fd_data in fds.fd_data.values_mut() {
+                    let PollFd(poll_fd) = &mut fds.poll_fds[fd_data.poll_fds_index];
+
+                    if pending_removal.contains(&poll_fd.fd) {
+                        // Fd was removed while poll was running
+                        continue;
+                    }
+
+                    if poll_fd.revents != 0 {
+                        // Store event
+                        events.push(Event {
+                            token: fd_data.token,
+                            events: poll_fd.revents,
+                        });
+
+                        if poll_fd.revents & (libc::POLLHUP | libc::POLLERR) != 0 {
+                            pending_removal.push(poll_fd.fd);
+                            closed_raw_fds.push(poll_fd.fd);
+                        }
+
+                        // Remove the interest which just got triggered
+                        // the IoSourceState/WakerRegistrar used with this selector will add back
+                        // the interest using reregister.
+                        poll_fd.events &= !poll_fd.revents;
+
+                        if events.len() == num_fd_events {
+                            break;
+                        }
+                    }
+                }
+
+                break;
+            }
+        }
+
+        drop(fds);
+        let _ = self.deregister_all(&closed_raw_fds);
+
+        Ok(())
+    }
+
+    pub fn register(&self, fd: RawFd, token: Token, interests: Interest) -> io::Result<()> {
+        self.register_internal(fd, token, interests).map(|_| ())
+    }
+
+    pub fn register_internal(
+        &self,
+        fd: RawFd,
+        token: Token,
+        interests: Interest,
+    ) -> io::Result<Arc<RegistrationRecord>> {
+        #[cfg(debug_assertions)]
+        if fd == self.notify_waker.as_raw_fd() {
+            return Err(io::Error::from(io::ErrorKind::InvalidInput));
+        }
+
+        // We must handle the unlikely case that the following order of operations happens:
+        //
+        // register(1 as RawFd)
+        // deregister(1 as RawFd)
+        // register(1 as RawFd)
+        // <poll happens>
+        //
+        // Fd's pending removal only get cleared when poll has been run. It is possible that
+        // between registering and deregistering and then _again_ registering the file descriptor
+        // poll never gets called, thus the fd stays stuck in the pending removal list.
+        //
+        // To avoid this scenario we remove an fd from pending removals when registering it.
+        let mut pending_removal = self.pending_removal.lock().unwrap();
+        if let Some(idx) = pending_removal.iter().position(|&pending| pending == fd) {
+            pending_removal.remove(idx);
```

Comment:
I don't think we care about the order, right?
```suggestion
            pending_removal.swap_remove(idx);
```